psl-core/src/main/java/org/linqs/psl/reasoner/sgd/SGDReasoner.java

```diff
@@ -60,60 +88,143 @@ public double optimize(TermStore baseTermStore) {
+                }
+
+                termCount++;
+                meanMovement += minimize(term, termStore.getVariableValues(), iteration);
             }
 
-            if (variableValues.length != 0) {
-                movement /= variableValues.length;
+            if (termCount != 0) {
+                meanMovement /= termCount;
             }
 
-            long end = System.currentTimeMillis();
+            if (breakOptimization(iteration, objective, oldObjective, meanMovement, termCount)) {
+                totalTime += System.currentTimeMillis() - start;
+                break;
+            }
 
-            oldObjective = objective;
-            objective = computeObjective(termStore);
+            // Keep track of the old variables for a deferred objective computation.
+            if (oldVariableValues == null) {
+                oldVariableValues = Arrays.copyOf(termStore.getVariableValues(), termStore.getVariableValues().length);
+                oldObjective = Double.POSITIVE_INFINITY;
+            } else {
+                System.arraycopy(termStore.getVariableValues(), 0, oldVariableValues, 0, oldVariableValues.length);
+                oldObjective = objective;
+            }
+
+            long end = System.currentTimeMillis();
             totalTime += end - start;
 
             if (log.isTraceEnabled()) {
-                log.trace("Iteration {} -- Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimiztion Time: {}",
-                        iteration, objective, movement, (end - start), totalTime);
+                log.trace("Iteration {} -- Objective: {}, Normalized Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimization Time: {}",
+                        iteration, objective, objective / termCount, meanMovement, (end - start), totalTime);
             }
 
             iteration++;
             termStore.iterationComplete();
-
-            if (breakOptimization(iteration, objective, oldObjective, movement)) {
-                break;
-            }
         }
 
-        log.info("Optimization completed in {} iterations. Objective: {}, Total Optimiztion Time: {}",
-                iteration - 1, objective, totalTime);
-        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termStore.size());
+        objective = computeObjective(termStore);
+        log.info("Optimization completed in {} iterations. Objective: {}, Normalized Objective: {}, Total Optimization Time: {}",
+                iteration, objective, objective / termCount, totalTime);
+        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termCount);
 
         termStore.syncAtoms();
 
         return objective;
     }
 
-    private boolean breakOptimization(int iteration, double objective, double oldObjective, float movement) {
+    private float minimize(SGDObjectiveTerm term, float[] variableValues, int iteration) {
```

Comment:
moving this out to reasoner is a bit strange. I see why you have done it but it makes things inconsistent. ADMMReasoner has all minimize in its term which I think is cleaner.